### PR TITLE
8 add module external_file_location

### PR DESCRIPTION
**Pull request by bealdav**
This module allow the exchange of files with main protocols ftp, sftp.

The purpose of this module is not to import the data of the file (only exchange files).

Example of kind files which can use this module:
- ecommerce: exchange with marketplace: ebay, amazon, etc.
- logistics: 
  - send delivery order to logistics center, receive a delivery reply, 
  - send a notification of goods send to this center and collect a confirmation of good receiving of goods by the logistics
  - collect inventory of the center
- banking: receiving all documents ...

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -22,17 +22,19 @@ env:
...
   system_site_packages: true
 
 install:
+  - git clone https://github.com/akretion/server-tools.git ${HOME}/server-tools-attachm -b 8-attachm-meta
   - git clone https://github.com/OCA/maintainer-quality-tools.git ${HOME}/maintainer-quality-tools
   - export PATH=${HOME}/maintainer-quality-tools/travis:${PATH}
   - travis_install_nightly
   - printf '[options]\n\nrunning_env = dev\n' > ${HOME}/.openerp_serverrc
   - ln -s ${TRAVIS_BUILD_DIR}/server_environment_files_sample ${TRAVIS_BUILD_DIR}/server_environment_files
+  - pip install fs
```
Comment (moylop260) [inline .travis.yml]: What about add `fs` in [requirements.txt](https://github.com/OCA/server-tools/blob/8.0/requirements.txt)?

File: external_file_location/__openerp__.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    'website': 'www.akretion.com',
+    'license': 'AGPL-3',
+    'category': 'Generic Modules',
+    'depends': [
+        'attachment_metadata',
+    ],
+    'external_dependencies': {
+        'python': [
+            'fs',
+            'paramiko',
```
Comment (moylop260) [inline external_file_location/__openerp__.py:19]: Should we add `paramiko` to requirements.txt too?

File: external_file_location/__openerp__.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    'data': [
+        'views/menu.xml',
+        'views/attachment_view.xml',
+        'views/location_view.xml',
+        'views/task_view.xml',
+        'data/cron.xml',
+        'security/ir.model.access.csv',
+    ],
+    'installable': True,
+    'application': True,
```
Comment (moylop260) [inline external_file_location/__openerp__.py]: Why is a `application`?

File: external_file_location/models/__init__.py [hunk 4]
```diff
@@ -0,0 +1,3 @@
+from . import attachment
```
Comment (moylop260) [inline external_file_location/models/__init__.py:1]: license?

File: oca_dependencies.txt [hunk 5]
```diff
@@ -0,0 +1 @@
+server-tools-file https://github.com/akretion/server-tools.git 8-attachm-meta
```
Comment (moylop260) [inline oca_dependencies.txt]: Is posible migrate it to a oca project?

Comment (mourad-ehm) [inline oca_dependencies.txt]: The branch 8-attachm-meta was proposed in this pull request :  
https://github.com/OCA/server-tools/pull/376

File: external_file_location/__openerp__.py (hunk 3 above)
Comment (mourad-ehm) [inline external_file_location/__openerp__.py]: It's not application, I will correct it.

File: .travis.yml [hunk 6]
```diff
@@ -22,17 +22,19 @@ env:
...
   - TESTS="1" ODOO_REPO="odoo/odoo"
   - TESTS="1" ODOO_REPO="OCA/OCB"
-  - UNIT_TEST="1" 
+  - UNIT_TEST="1"
 
 virtualenv:
   system_site_packages: true
 
 install:
+  - git clone https://github.com/akretion/server-tools.git ${HOME}/server-tools-attachm -b 8-attachm-meta
```
Comment (hparfr) [inline .travis.yml]: to be removed

File: oca_dependencies.txt (hunk 5 above)
Comment (moylop260) [inline oca_dependencies.txt]: <s>I saw that the module `attachment_metadata` is dependency just to add fields to xml view. (Fix me if I'm wrong.)
What about use directly the original view and remove this dependency to allow choose each user install optionally `attachment_metadata`?</s>

Comment (moylop260) [inline oca_dependencies.txt]: Forget it, I saw that is used the model too.

**Comment by lmignon** [comment]
@bealdav The same features are provided by https://github.com/OCA/connector-interfaces/pull/17
This PR is the result of https://github.com/OCA/connector-interfaces/issues/5

**Comment by bealdav** [comment]
The complete story is here: https://github.com/initOS/connector-interfaces/pull/1
This module is an evolution of file exchange to be compatible with connector interfaces.

**Comment by lmignon** [comment]
what about OCA/connector-interfaces#17 ?

**Comment by sebastienbeau** [comment]
Hi @lmignon 
In fact, this module do not have the same behaviour. This module is just a piece of the connector flow tool. We modify our work to be compatible with this PR.
If you look here 
- https://github.com/initOS/connector-interfaces/pull/1
- https://github.com/OCA/connector-interfaces/pull/17#discussion_r32612829
  Still waiting for some feedback

The aim is to split the feature in small part
- having a tool to exchange File (this PR)
- having an ETL that use this module : connector-flow
- having extra tool (amazon connector, ebay, bank statement...) that use also the exchange of file

**Comment by bealdav** [comment]
@lmignon I think it's important than features are extracted in small modules, don't you think ? tests, and migration easier, more reusable.

**Comment by lmignon** [comment]
> @lmignon I think it's important than features are extracted in small modules, don't you think ? tests, and migration easier, more reusable.

@bealdav  For sure :smirk: 

@sebastienbeau Thank you for the explanation.

On a first read the code looks really good! I need to find times to review it a little deeper.

File: external_file_location/__openerp__.py (hunk 2 above)
Comment (lasley) [inline external_file_location/__openerp__.py:19]: I don't see it being used in the code actually. Looks like the depends are now `ftplib` and `pysftp`

File: external_file_location/tasks/abstract_fs.py [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+    _name = None
+    _key = None
+    _synchronize_type = None
+    _default_port = None
+
+    def __init__(self, env, config):
+        self.env = env
+        self.host = config.get('host', '')
+        self.user = config.get('user', '')
+        self.pwd = config.get('pwd', '')
```
Comment (lasley) [inline external_file_location/tasks/abstract_fs.py:61]: Is it possible to allow key auth as well?

**Pull request closed without merging**